pallets/xyk/src/lib.rs

```diff
@@ -948,6 +875,220 @@ impl<T: Config> AMM<T::AccountId, AssetId, AssetPair, Balance> for Pallet<T> {
+		ensure!(Self::exists(pair), ExecutorError::Error(()));
+
+		let pair_account = Self::get_pair_id(pair);
+
+		let asset_in_reserve = T::Currency::free_balance(asset_in, &pair_account);
+		let asset_out_reserve = T::Currency::free_balance(asset_out, &pair_account);
+
+		let amount_in_without_fee = amount_in
+			.amount
+			.checked_sub(amount_in.fee)
+			.ok_or(ExecutorError::Error(()))?;
+
+		let amount_out =
+			hydra_dx_math::xyk::calculate_out_given_in(asset_in_reserve, asset_out_reserve, amount_in_without_fee)
+				.map_err(|_| ExecutorError::Error(()))?;
+
+		let transfer_fee = Self::calculate_fee(amount_out).map_err(|_| ExecutorError::Error(()))?;
+
+		Ok(TradeCalculation {
+			amount: amount_out,
+			fee: transfer_fee,
+		})
+	}
+
+	fn calculate_buy(
+		pool_type: PoolType<AssetId>,
+		asset_in: AssetId,
+		asset_out: AssetId,
+		amount_out: Self::TradeCalculationResult,
+	) -> Result<Self::TradeCalculationResult, ExecutorError<Self::Error>> {
+		if pool_type != PoolType::XYK {
+			return Err(ExecutorError::NotSupported);
+		}
+
+		let pair = AssetPair { asset_in, asset_out };
+		ensure!(Self::exists(pair), ExecutorError::Error(()));
+
+		let pair_account = Self::get_pair_id(pair);
+
+		let asset_in_reserve = T::Currency::free_balance(asset_in, &pair_account);
+		let asset_out_reserve = T::Currency::free_balance(asset_out, &pair_account);
+
+		let amount_out_with_fee = amount_out
+			.amount
+			.checked_add(amount_out.fee)
+			.ok_or(ExecutorError::Error(()))?;
+
+		let amount_in =
+			hydra_dx_math::xyk::calculate_in_given_out(asset_out_reserve, asset_in_reserve, amount_out_with_fee)
+				.map_err(|_| ExecutorError::Error(()))?;
+
+		let transfer_fee = Self::calculate_fee(amount_in).map_err(|_| ExecutorError::Error(()))?;
+
+		Ok(TradeCalculation {
+			amount: amount_in,
+			fee: transfer_fee,
+		})
+	}
+
+	fn execute_sell(
+		pool_type: PoolType<AssetId>,
+		who: &T::AccountId,
+		asset_in: AssetId,
+		asset_out: AssetId,
+		amount_in: Self::TradeCalculationResult,
+	) -> Result<(), ExecutorError<Self::Error>> {
+		if pool_type != PoolType::XYK {
+			return Err(ExecutorError::NotSupported);
+		}
+
+		let amount_out = Self::calculate_sell(pool_type, asset_in, asset_out, amount_in)?;
+
+		let assets = AssetPair { asset_in, asset_out };
+		let fee = (asset_out, amount_out.fee);
+		let amount_in_without_fee = amount_in
+			.amount
+			.checked_sub(amount_in.fee)
+			.ok_or(ExecutorError::Error(()))?;
+		let amount_out_without_fee = amount_out
+			.amount
+			.checked_sub(amount_out.fee)
+			.ok_or(ExecutorError::Error(()))?;
+
+		let amm_transfer: AMMTransfer<T::AccountId, AssetId, AssetPair, Balance> = AMMTransfer {
+			origin: who.clone(),
+			assets,
+			amount: amount_in_without_fee,
+			amount_out: amount_out_without_fee,
+			fee,
+			discount: false,
+			discount_amount: 0,
+		};
+		//TODO: Dani - proper error propagation
+		Self::do_validate_sell_common(amount_in_without_fee, assets, &who).map_err(|_| ExecutorError::Error(()))?;
```

Comment:
just let it use its own error, those errors should be propagated to the executor as well, exectutor shouldn't need any generice 'TradeFailed' error

Reply:
you are right, let's retain the original error info, there is indeed no need for additional abstraction around them.

On the other hand, I need to keep `ExecutorError` as it is a part of the control flow. 

I wonder what other things I could do for example some random calculation error. Since we want to return/propagate `DispatchError,` (wrapped in ExcecutorError) I would need to use `DispatchError::Other("...")`, so for example:
```
        let amount_in_without_fee = amount_in
            .amount
            .checked_sub(amount_in.fee)
            .ok_or(ExecutorError::Error(DispatchError::Other("Error while calculating sell trade")))?;
````

Is DisptachError::Other a good option here?

Reply:
you are right, let's retain the original error info, there is indeed no need for additional abstraction around them.

On the other hand, I need to keep `ExecutorError` as it is a part of the control flow. 

I wonder what other things I could do for example some random calculation error. Since we want to return/propagate `DispatchError,` (wrapped in ExcecutorError) I would need to use `DispatchError::Other("...")`, so for example:
```
        let amount_in_without_fee = amount_in
            .amount
            .checked_sub(amount_in.fee)
            .ok_or(ExecutorError::Error(DispatchError::Other("Error while calculating sell trade")))?;
````

Is DisptachError::Other a good option here?

Reply:
why you need `ExecutorError` for control flow? any error should abort the execution

Reply:
as discussed in the mob-programming session, I leave the error as it is used in the control flow. I mark the comment resolved.